File: Runtime/MRSmokeTestMain.cs
//----------------------------------------------------------------------------
//
//  $Id: MRSmokeTestMain.cs 470 2009-06-12 17:43:02Z hoffman $
//
//  $Author: hoffman $
//  $Date: 2009-06-12 13:43:02 -0400 (Fri, 12 Jun 2009) $
//  $Revision: 470 $
//
//  Copyright (C) 2007 Kitware, Inc.
//
//----------------------------------------------------------------------------

//----------------------------------------------------------------------------
namespace Kitware.mummy.Runtime
{
  /// <summary>
  /// </summary>
  public class SmokeTestMain
  {
    /// <summary>
    /// </summary>
    public static void Main(string[] args)
    {
      int i = 0;
      int n = 10000;
      bool strong = false;
      bool evensplit = false;
      bool pseudorandom = false;

      if (args.Length>0)
      {
        n = System.Convert.ToInt32(args[0]);
      }

      if (args.Length>1)
      {
        if ("--strong" == args[1])
        {
          strong = true;
        }
        else if ("--evensplit" == args[1])
        {
          evensplit = true;
        }
        else if ("--pseudorandom" == args[1])
        {
          pseudorandom = true;
        }
      }

      Kitware.mummy.Runtime.MRSmokeTest test = new Kitware.mummy.Runtime.MRSmokeTest(strong);
      Kitware.mummy.Runtime.MRSmokeTest testClone = test.Clone();

      Kitware.mummy.Runtime.Methods.Trace("looping...");

      for (i= 0; i<args.Length; ++i)
      {
        Kitware.mummy.Runtime.Methods.Trace(System.String.Format("arg{0}: {1}", i, args[i]));
      }

      System.Random r = new System.Random();

      for (i= 0; i<n; ++i)
      {
        if (evensplit)
        {
          strong = !strong;
        }
        else if (pseudorandom)
        {
          strong = (r.NextDouble() > 0.5 ? true : false);
        }

        test = new Kitware.mummy.Runtime.MRSmokeTest(strong);
      }

      if (null == testClone)
      {
        throw new System.Exception("error: testClone is null!");
      }

      test = null;
      testClone = null;

      uint teCount = Kitware.mummy.Runtime.Methods.GetTypeEntryCount();
      Kitware.mummy.Runtime.Methods.Trace(System.String.Format("teCount: {0}", teCount));

      Kitware.mummy.Runtime.Methods.Trace("about to call System.GC.Collect()...");

      System.GC.Collect();

      Kitware.mummy.Runtime.Methods.Trace("done with System.GC.Collect()...");
    }
  }
}

File: Runtime/MRSmokeTestMain.cxx
//----------------------------------------------------------------------------
//
//  $Id: MRSmokeTestMain.cxx 418 2009-03-19 19:41:27Z hoffman $
//
//  $Author: hoffman $
//  $Date: 2009-03-19 15:41:27 -0400 (Thu, 19 Mar 2009) $
//  $Revision: 418 $
//
//  Copyright (C) 2007 Kitware, Inc.
//
//----------------------------------------------------------------------------

#include "MRSmokeTest.h"
#include "MummyRuntime.h"
#include <iostream>
#include "stdlib.h" // atoi

//----------------------------------------------------------------------------
int main(int argc, char* argv[])
{
  int err = 0;
  int i = 0;
  int n = 10000;

  if (argc>1)
  {
    n = atoi(argv[1]);
  }

  // Call ObjectRefCountChanged with bogus parameters (NULL
  // pointer, silly ref counts) to make sure it's robust...
  //
  Kitware::mummy::Runtime::ObjectRefCountChanged(0, 333, 444);

  // Call Verify *before* the loop:
  //
  std::cout << "Calling Kitware::mummy::Runtime::Verify" << std::endl;
  Kitware::mummy::Runtime::Verify();

  std::cout << "n: " << n << std::endl;
  for (i= 0; i<n; ++i)
  {
    MRSmokeTest* test = new MRSmokeTest;
    MRSmokeTest* clone = test->Clone();
    Kitware::mummy::Runtime::LookupInstanceTypeEntry(test);
    Kitware::mummy::Runtime::LookupInstanceTypeEntry(clone);
    delete clone;
    delete test;
  }

  // Valid calls (as long as there is at least one entry by now...):
  //
  std::cout << "Testing mummy::Runtime methods with valid calls..." << std::endl;
  MRSmokeTest* test2 = new MRSmokeTest;
  MRSmokeTest* clone2 = test2->Clone();

  Kitware::mummy::Runtime::ObjectRefCountChanged(test2, 1, 2);
  Kitware::mummy::Runtime::LookupInstanceTypeEntry(test2);
  Kitware::mummy::Runtime::LookupInstanceTypeEntry(clone2);
  Kitware::mummy::Runtime::GetTypeEntryByIndex(0);
  Kitware::mummy::Runtime::GetTypeEntryClassNameKey(0);
  Kitware::mummy::Runtime::ObjectRefCountChanged(test2, 2, 1);

  delete clone2;
  delete test2;

  std::cout << "Calling Kitware::mummy::Runtime::Print" << std::endl;
  std::cout << Kitware::mummy::Runtime::Print() << std::endl;

  std::cout << "Calling Kitware::mummy::Runtime::Verify" << std::endl;
  Kitware::mummy::Runtime::Verify();

  std::cout << "Calling Kitware::mummy::Runtime::GetTypeEntryByIndex(0)" << std::endl;
  Kitware::mummy::TypeEntry* entry = Kitware::mummy::Runtime::GetTypeEntryByIndex(0);
  if (entry)
    {
    std::cout << "entry:" << std::endl;
    std::cout << entry->Print() << std::endl;
    entry->GetObjectRefCountChangedCallback();
    }
  else
    {
    std::cout << "error: no entry found at index 0" << std::endl;
    err = 1;
    }

  // Invalid calls (unless there are somehow 4 billion++ entries...):
  //
  std::cout << "Testing mummy::Runtime methods with invalid calls..." << std::endl;
  Kitware::mummy::Runtime::ObjectRefCountChanged(0, 0, 0);
  Kitware::mummy::Runtime::LookupInstanceTypeEntry(0);
  Kitware::mummy::Runtime::GetTypeEntryByIndex((Kitware::mummy::Runtime::TypeEntryIndex) -1);
  Kitware::mummy::Runtime::GetTypeEntryClassNameKey((Kitware::mummy::Runtime::TypeEntryIndex) -1);

  std::cout << "Test done" << std::endl;

  return err;
}
